example/lib/ide/problems_panel/problems_panel.dart

```diff
@@ -0,0 +1,78 @@
+import 'dart:io';
+
+import 'package:example/ide/problems_panel/diagnostics.dart';
+import 'package:example/lsp_exploration/lsp/lsp_client.dart';
+import 'package:flutter/material.dart';
+
+class ProblemsPanel extends StatefulWidget {
+  const ProblemsPanel({
+    super.key,
+    required this.lspClient,
+    this.sourceFile,
+  });
+
+  final LspClient lspClient;
+  final File? sourceFile;
+
+  @override
+  State<ProblemsPanel> createState() => _ProblemsPanelState();
+}
+
+class _ProblemsPanelState extends State<ProblemsPanel> {
+  List<Diagnostic>? diagnostics;
+  String? uri;
+
+  @override
+  void initState() {
+    super.initState();
+    widget.lspClient.addNotificationListener(_onLspNotification);
```

Comment:
We should be implementing `didUpdateWidget` and  `dispose` to remove the listeners.

Reply:
Can you explain the `didUpdateWidget` lifecycle relationship to removing the listener? In `_IDEState` there is no `didUpdateWidget` override either, so I didn't have anything to go on. Are you saying that if the widget gets updated in a certain way, that it is possible to lose the reference to the listener?

Reply:
Whenever we add a listener to an object received from the widget, we need to account for the fact that in a subsequent build it's possible that we receive a different object.

In our example, it's possible that the object that creates the `ProblemsPanel` builds it with a different `lspClient`. Because of that, we need to remove the listener from the old object and add it to the new one. The pattern will be always:

```dart
@override
void didUpdateWidget(ClassOfTheWidget oldWidget) {
  super.didUpdateWidget(oldWidget);
  if (widget.myListenable != oldWidget.myListenable) {
    oldWidget.myListenable.removeListener(_myListener);
    widget.myListenable.addListener(_myListener);
  }
}
```

So in our case we would have:

```dart
@override
void didUpdateWidget(ProblemsPanel oldWidget) {
  super.didUpdateWidget(oldWidget);
  if (widget.lspClient != oldWidget.lspClient) {
    oldWidget.lspClient.removeNotificationListener(_onLspNotification);
    widget.lspClient.addNotificationListener(_onLspNotification);
  }
}
```

Reply:
That makes sense. Thank you. The part I was missing was receiving the object from the widget rather than creating the object in a stateful widget.